File: 92.ReverseLinkedListII/reverseLinkedListII.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[listNode], left: int, right: int) -> Optional[ListNode]:
        if not head or left == right:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        
        # Move prev to the node just before the left position
        for _ in range(left - 1):
            prev = prev.next
        
        # Start reversing from the left position
        current = prev.next
        tail = current
        
        for _ in range(right - left + 1):
            next_node = current.next
            current.next = prev.next
            prev.next = current
            current = next_node
        
        tail.next = current
        
        return dummy.next

File: 92.ReverseLinkedListII/reverseLinkedListII.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 * 
 * @param head (ListNode | null)
 * @param left (number)
 * @param right (number)
 * @returns (ListNode | null)
 */

function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
    if (!head || left === right) return head;

    let dummy = new ListNode(0, head);
    let prev = dummy;

    // Move `prev` to the node just before the `left` position
    for (let i = 1; i < left; i++) {
        prev = prev.next!;
    }

    let current = prev.next;
    let next: ListNode | null = null;

    // Reverse the sublist from `left` to `right`
    for (let i = 0; i < right - left; i++) {
        next = current!.next!;
        current!.next = next.next;
        next.next = prev.next;
        prev.next = next;
    }

    return dummy.next;
}

